# Extra Information

### Principal Component Analysis (PCA) plot

The PCA plot can show whether your samples separate into groups, what taxa (or groups of tax) are driving this separation, and what taxa are irrelevant for your analysis. More information about how to interpret PCA plots and how they can be used can be found [here](https://github.com/ggloor/CoDa_microbiome_tutorial/blob/master/first_biplot.pdf).

#### Filtering options

There are several ways you can filter your data before a biplot is made. The effect of the first slider ('Read Cutoff') is changed by which button is selected. Presence (or absence) of the taxonomy column does not influence the filtering.

1. **Minimum count per OTU**  
This filter removes any rows with a maximum count with less than the filter.
  
2. **Minimum count sum per sample**  
This filter removes any columns that have a lower total sum of counts than the filter.  

3. **Minimum proportional abundance**  
This filter calculates frequency for each sample, and removes any rows that have no frequencies above the threshold (ie, all samples for one OTU have abundance less than 0.01).

4. **Maximum proportional abundance**  
This filter calculates frequency for each sample, and removes any rows that have no frequencies below the threshold (ie, all samples for one OTU have abundance less than 0.01).

5. **Minimum count sum per OTU**  
This filter removes any rows that have a lower total sum of counts than the filter.

**Variance cutoff**

This filter calculates the variance for each OTU (row), and removes any rows that have lower variance than the filter. Variance is calculated after the reads have been transformed by the centre-log ratio. 

**Adjust scale**

This changes the scale of the biplot. When set to zero, it shows the relationship between samples (columns), while being set to 1 shows the relationship between OTUs (rows). Data is not filtered from this operation.

### Association Plot

Association plots can show positively correlated taxa, where the ratio between two or more taxa is approximately constant in all samples. The *phi* metric is used, which can be changed depending on the strength of assocation you wish to display (more information about the *phi* metric can be found from [Lovell et al. (2015).](https://doi.org/10.1371/journal.pcbi.1004075) A guide to your first association plot can be found [here.](https://github.com/ggloor/CoDa_microbiome_tutorial/blob/master/first_association.pdf)

### Relative Abundance Plots

A dendrogram and stacked barplot calculates the sum of counts based on genera, showing the relative abundance of each genera for each sample, colour coded by taxonomy. For this to work, you must have a taxonomy column that has at least 6 taxonomic levels. Different method of calculating the distance matrix and .Information about the different clustering methods are available under the details section [of this article.](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/hclust.html)

### Effect plots

Effect plots are useful at showing the size of effect for between group differences. It differs from a Volcano plot, since the effect size is plotted rather than the p-value, giving information about the size of the effect. [Gloor et al. 2016](http://dx.doi.org/10.1080/10618600.2015.1131161) details the differences between effect plots and volcano plots, and what information you can actually obtain from them. 

By hovering your mouse of points on the plot, the sample ID and information will be displayed. 

### Example data

Two published example datasets are available to use in omicplotR. The vaginal dataset contains data and metadata, published by [Macklaim et al. (2015).](http://www.tandfonline.com/doi/full/10.3402/mehd.v26.27799) 

The 'selex' dataset contains a dataset without metadata, published by [McMurrough et al. (2014)](http://www.pnas.org/content/111/23/E2376.abstract). When generating effect plots using this data, the first 7 columns are selective conditions and the next 7 columns are grown under non-selective conditions (so enter both conditions manually as 7 columns).










